whd/dispatcher.py

```diff
@@ -318,6 +318,56 @@ def is_up_to_date(resource, resource_versions):
             sleep_seconds=sleep_seconds*1.2,
         )
 
+    def _jobs_not_triggered(self, concourse_api, matching_resources, pr_event):
+        '''
+        return all jobs and the corresponding resource versions, which have not been triggered
+        '''
+        for resource in matching_resources:
+            pipeline_name = resource.pipeline.name
+            resource_versions = concourse_api.resource_versions(
+                pipeline_name=pipeline_name,
+                resource_name=resource.name
+            )
+            pr_resource_versions = [
+                rv for rv in resource_versions if rv.version()['pr'] == str(pr_event.number())
+            ]
+            if not pr_resource_versions:
+                logger.warning(
+                    f"no resource versions found for pipeline {resource.pipeline.name} "
+                    f"resource {resource.name} and PR {pr_event.number()}"
+                )
+                return
+
+            # get all jobs which get triggered by the resource
+            jobs_triggered_by_resource = [
+                job for job in resource.pipeline.jobs if job.is_triggered_by_resource(resource.name)
+            ]
+
+            # check if there is a pr resource version which has not triggered a job.
+            # If so return the job and the resource version
+            for job in jobs_triggered_by_resource:
+                builds = concourse_api.job_builds(pipeline_name, job.name)
```

Comment:
again, I would figure this filtering should rather be moved into the api class, or moved to a util function. please do not make the dispatcher too large. the implementation details of how to tell whether or not a job was already triggered should not be this class's concern

Reply:
will move all "pinning" code to own module